### ISSUE TITLE
XL: client timeout during ListObjects - May 27th

### ISSUE THREAD

**Issue by krishnasrinivas** [minio/minio#1516]
In ListDir, at XL layer, for each entry we do a Stat() call to see if it's a leaf directory.
At object-layer, for non-recursive case we again do a Stat() to check if it's a multipart uploaded file. This will cause the client to timeout before it receives first batch of 1000 entries.

**Comment by harshavardhana** [comment]
> In ListDir, at XL layer, for each entry we do a Stat() call to see if it's a leaf directory.
> At object-layer, for non-recursive case we again do a Stat() to check if it's a multipart uploaded file. This will cause the client to timeout before it receives first batch of 1000 entries.

ON which server are you observing this? - can you paste 'dd' results?

**Comment by krishnasrinivas** [comment]
@harshavardhana this is the issue we discussed yesterday for list ListObjects, not GetObject. Potential fix is to use `.__leaf__` (predictable) suffix to avoid a stat() to see if it's a leaf dir.

This is not a bug seen as of now, but potential issue based on previous experience with listobjects.

**Comment by harshavardhana** [comment]
> @harshavardhana this is the issue we discussed yesterday for list ListObjects, not GetObject. Potential fix is to use .**leaf** (predictable) suffix to avoid a stat() to see if it's a leaf dir.
> 
> This is not a bug seen as of now, but potential issue based on previous experience with listobjects.

I think we shouldn't speculate here, we still need to reproduce a valid case here.  See the breaking point where exactly we see this in what conditions.

**Comment by krishnasrinivas** [comment]
Sure I can reproduce it, but just wanted to put it as an issue so that we don't forget about it as we had decided to open an issue if we think there is an issue. We can close this if it's not reproducible.

**Comment by harshavardhana** [comment]
> Sure I can reproduce it, but just wanted to put it as an issue so that we don't forget about it as we had decided to open an issue if we think there is an issue. We can close this if it's not reproducible.

yeah makes sense.

**Comment by harshavardhana** [comment]
Current code on .. minio2 server 

```
$ time mc ls myminio/testbucket/ -r | wc -l
1640

real    0m0.701s
user    0m0.096s
sys     0m0.020s
```

**Comment by harshavardhana** [comment]
Some profiling output on what is going on with ListObjects

[Current master with no optimization](https://github.com/minio/minio/files/256176/list-cpu-noopt.pdf)

[With optimization at getPartsMetadata](https://github.com/minio/minio/files/256177/list-cpu-opt.pdf)

There is a drop of around from 450 secs to 316 secs on the overall call flow, both tests were run with "echo "3" drop_caches" so they all are all fresh run fetched from the disk. 

Sending a patch.

**Comment by harshavardhana** [comment]
> Some profiling output on what is going on with ListObjects
> 
> Current master with no optimization
> 
> With optimization at getPartsMetadata
> 
> There is a drop of around from 450 secs to 316 secs on the overall call flow, both tests were run with "echo "3" drop_caches" so they all are all fresh run fetched from the disk.
> 
> Sending a patch.

Latter CPU usage is cumulative listing of 165k files. 

```
$ mc ls -r myminio/testbucket/ | wc -l
165660
```

**Comment by harshavardhana** [comment]
@krishnasrinivas is this reproducible still ?

**Comment by krishnasrinivas** [comment]
@harshavardhana it should stil be reproducible. I will check today.

**Comment by harshavardhana** [comment]
> @harshavardhana it should stil be reproducible. I will check today.

Can you give me access to the setup? need to get the blocking profile and cpu profile.

**Comment by krishnasrinivas** [comment]
will do

**Comment by krishnasrinivas** [comment]
the problem is still there for slow disks (azure vm with 30mbps sequential writes)

```
minio@minio2:~$ time mc ls myminio/test1/onemil/|head
mc: <ERROR> Unable to stat ‘myminio/test1/onemil/’. Get http://127.0.0.1:9000/test1/?delimiter=%2F&max-keys=1000&prefix=onemil%2F: net/http: request canceled (Client.Timeout exceeded while awaiting headers)

real    16m0.072s
user    0m0.012s
sys     0m0.008s
minio@minio2:~$
```

could not get blocking/cpu profile because of https://github.com/minio/minio/pull/1670
Will give you access.

**Comment by harshavardhana** [comment]
[list-objects-blocking.pdf](https://github.com/minio/minio/files/274142/list-objects-blocking.pdf)
[list-objects-cpu.pdf](https://github.com/minio/minio/files/274143/list-objects-cpu.pdf)

Attaching here for reference.

**Comment by harshavardhana** [comment]
> list-objects-blocking.pdf
> list-objects-cpu.pdf
> 
> Attaching here for reference.

@krishnasrinivas were you able to validate with attaching addition disks from 3 to 12 ?

**Comment by harshavardhana** [comment]
Assigning to @krishnasrinivas

**Comment by krisis** [comment]
The following are the steps I performed and I am attaching my observations so far.
- Applied the patch (found here: `play/0001-Used-getObjectinfo-instead-of-Stat-for-perf-measurem.patch`) to use `xl.getObjectInfo` function to determine if an entry is an object or a prefix instead of `xl.isObject` (which performs an `os.Stat` call).
- `mc ls` on 1M object entries with and without the patch took unreasonably long times and I captured a cpu profile for both the cases.

[1] - `play/getObjectInfoVsStat.pdf` - with the above patch applied
[2] - `play/vanilla-cpu-mc-ls.pdf` - without the patch based on `rewrite-xl^HEAD`
